Fixes (partially): oppia can not not run on Windows

cuichenli opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
So with some changes, I have made the project to be runnable on Windows. 
Known issues:
- pre_commit_linter will break if there are js files to lint. Due to the `Script` object generated by `esprima` module can not be pickled. Which cause some errors on multiprocessing on Windows (thanks to myself).
- All the github action builds (windows, macos, ubuntu) are failing on one backend test. Error message: 
```
WARNING: FAILED TO RUN core.domain.exp_services_test.LoadingAndDeletionOfExplorationDemosTests

This is most likely due to an import error.
```
Build link: https://github.com/cuichenli/oppia/runs/297434057
- Currently, a lot of backend tests are broken on Windows. This should be fixed later. 
## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
Dirty tweak

Commit:
Fix pre_commit_linter

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Fix backend test

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Fix lint issue

Commit:
Fix issue with node and yarn

Commit:
Fix node download

Commit:
Try github action

Commit:
Use with python

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update install_third_party_libs.py

cuichenli committed:
Update install_third_party_libs.py

cuichenli committed:
Update install_third_party_libs.py

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

Commit:
Make some unit test compatible with windows

Commit:
Add continue on error

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Fix linter issues

Commit:
Fix accidentely deleted line

Commit:
Fix unit tests

Commit:
Remove linter step from github action

Commit:
Fix

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Test

cuichenli committed:
Update build.yml

Commit:
Track

Commit:
Revert back

Commit:
exclude load tests

Commit:
Update github action

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

cuichenli committed:
Update build.yml

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

kevinlee12 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,95 @@
+name: Backend Test
```
Don't commit this file yet :)

cuichenli replied:
Removed

kevinlee12 commented on scripts/pre_commit_linter.py:
```diff
@@ -1297,129 +1297,6 @@ def check_for_important_patterns_at_bottom_of_codeowners(important_patterns):
     return failed
 
 
-def _check_codeowner_file(verbose_mode_enabled):
```
Why was this removed?

cuichenli replied:
There are two `_check_codeowner_file` method defined in this file, so I just deleted one of them
https://github.com/oppia/oppia/blob/cb99e892f96b3bece25ad6cb409ee4ba5c4abe05/scripts/pre_commit_linter.py#L1300
https://github.com/oppia/oppia/blob/cb99e892f96b3bece25ad6cb409ee4ba5c4abe05/scripts/pre_commit_linter.py#L1300

kevinlee12 reviewed (commented):
thanks @cuichenli! I have two comments. ptal!

@bansalnitish can you run this on your mac to see if it works (it works on my Linux machine)? best to start from scratch since most of this affects setup. thanks!

Commit:
Remove github action file for now.

kevinlee12 commented:
@cuichenli backend tests coverage dropped below 100%, ptal!

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Enable codecoverage for Windows

Commit:
Add unit test for setup.py

Commit:
Add unit test for install_third_party_libs

Commit:
Fix lint issues

kevinlee12 commented on scripts/install_third_party_libs.py:
```diff
@@ -209,6 +209,21 @@ def ensure_pip_library_is_installed(package, version, path):
+        yarn_command = 'yarn.cmd'
+    else:
+        yarn_command = 'yarn'
+    return yarn_command
```
change this section to:
```python
    return 'yarn.cmd'
return 'yarn'
```

cuichenli replied:
Updated

kevinlee12 commented on scripts/run_backend_tests.py:
```diff
@@ -343,6 +350,11 @@ def main(args=None):
+        # Specify the path to coverage so it can be viewed as a module.
+        SUBPROCESS_ENV.update({
+            'PYTHONPATH': COVERAGE_DIR.encode(encoding='utf-8')
+                          # In case of some encoding issues on Windows.
```
Move this comment above line 354.

cuichenli replied:
Updated

kevinlee12 commented on scripts/setup_test.py:
```diff
@@ -143,6 +143,33 @@ def mock_print(msg_list):
+                'Here are some helpful articles:',
+                'http://docs.python-guide.org/en/latest/starting/install/win/',
+                'https://stackoverflow.com/questions/3701646/how-to-add-to-the-'
+                'pythonpath-in-windows-7'])
```
Move the contents of line 164 to line 163 and move the overflow to the next line. We want to make sure looks like an entire block for readability.

cuichenli replied:
I think that move L164 to L163 will actually make it unreadable since it is a list:

```
'It looks like you are using Windows. If you have Python '
'installed,', 'make sure it is in your PATH and that PYTHONPATH is set.',
'If you have two versions of Python (ie, Python 2.7 and 3), specify 2.7 '
'before other versions of Python when setting the PATH.',       
```

kevinlee12 replied:
What is there is fine. I think my comment went on an old commit.

kevinlee12 commented on scripts/setup_test.py:
```diff
@@ -338,6 +365,68 @@ def mock_getuid():
+            self.assertTrue(self.check_function_calls[item])
+        self.assertEqual(
+            self.urls, [
+                'https://nodejs.org/dist/v10.15.3/node-v10.15.3'
```
Please also make a comment in the setup script to change this string if the node version is upgraded.

cuichenli replied:
I made the node version and yarn version as global variables and share it with `setup_test.py`, this should solve the problem.

kevinlee12 commented on scripts/setup_test.py:
```diff
@@ -338,6 +365,68 @@ def mock_getuid():
+                'https://nodejs.org/dist/v10.15.3/node-v10.15.3'
+                '-win-x64.zip',
+                'https://github.com/yarnpkg/yarn/releases/download/'
+                'v1.17.3/yarn-v1.17.3.tar.gz'])
```
Ditto

kevinlee12 commented on scripts/setup_test.py:
```diff
@@ -338,6 +365,68 @@ def mock_getuid():
+                'https://nodejs.org/dist/v10.15.3/node-v10.15.3'
+                '-win-x86.zip',
+                'https://github.com/yarnpkg/yarn/releases/download/'
+                'v1.17.3/yarn-v1.17.3.tar.gz'])
```
Ditto

kevinlee12 reviewed (changes requested):
@cuichenli, I've left some comments, ptal!

Commit:
Fix issues

Simplify return statements
Share node version and yarn version between setup and setup_test

Commit:
Fix codecoverage

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

kevinlee12 commented on scripts/install_third_party_libs.py:
```diff
@@ -210,6 +209,20 @@ def ensure_pip_library_is_installed(package, version, path):
+                os.path.join(common.YARN_PATH, 'bin', 'yarn.sh'))
+        return 'yarn.cmd'
+    else:
+        return 'yarn'
```
You can get rid of `else`.

kevinlee12 reviewed (approved):
lgtm, just a nit!

kevinlee12 commented:
@vojtechjelinek @nithusha21 code owner review please!

nithusha21 commented on core/tests/test_utils.py:
```diff
@@ -1758,12 +1758,15 @@ def swap(self, obj, attr, newvalue):
         raise a RuntimeError.
         """
-        original = getattr(obj, attr)
+        original = getattr(obj, attr, 'NO SUCH ATTRIBUTE')
```
Should we make this string a global constant?

cuichenli replied:
Updated

nithusha21 reviewed (approved):
LGTM for codeowner files.

Commit:
Remove redundant else statement

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Add global variable

kevinlee12 reviewed (approved):
lgtm!

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

Commit:
Merge branch 'develop' of github.com:oppia/oppia into wip-run-on-windows

DubeySandeep commented on scripts/pre_commit_hook.py:
```diff
@@ -67,20 +68,23 @@ def install_hook():
 
     python_utils.PRINT('Making pre-commit hook file executable ...')
-    _, err_chmod_cmd = start_subprocess_for_result(chmod_cmd)
+    if platform.uname()[0] != 'Windows':
```
Why don't we want this to run on `windows` system? (Maybe we should add a comment on the top?)

cuichenli replied:
Windows system does not need these commands to make it executable.

DubeySandeep replied:
Please add a comment on top of this line. (It will be helpful for other contributors.) Also, if this check is done in multiple files then maybe we should have an `is_windows_os()` or something around that to avoid using the same hardcoded string in multiple places and the def name makes it more clear.)

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -63,8 +63,9 @@
-install_third_party_libs.main(args=[])
+from scripts import install_third_party_libs
+from scripts import build # isort:skip
+from scripts import docstrings_checker  # isort:skip
```
Can we add this in sorted order now?

DubeySandeep replied:
Can we add this in sorted order now?

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -595,17 +596,16 @@
 # pylint: enable=wrong-import-order
 # pylint: enable=wrong-import-position
 
+
```
Do we need this change?

cuichenli replied:
Do you mean the new line?

DubeySandeep replied:
Yes!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1986,12 +1869,15 @@ def _check_bad_patterns(self):
     def _check_patterns(self):
         """Run checks relate to bad patterns."""
-        methods = [self._check_bad_patterns, self._check_mandatory_patterns]
+        args = ((_FILES['.js'] + _FILES['.ts']), [], self.process_manager)
```
Why are we checking patterns only in `.js` and `.ts` files? (not .py, html and css files?)

Also, do we need to pass the `self.process_manager` as it's part of the same object?

cuichenli replied:
The way multiprocessing works in Windows requires the variables to be explicitly passed to the child process. https://stackoverflow.com/a/6596695/7676306

DubeySandeep replied:
Thanks for the link! :)

> Why are we checking patterns only in .js and .ts files?

I think you missed this one!

cuichenli replied:
This should lint all files 😢 . I am fixing it now.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1986,12 +1869,15 @@ def _check_bad_patterns(self):
-        self.process_manager['bad_pattern'] = summary_messages
-        _STDOUT_LIST.append(stdout)
+        process_manager['bad_pattern'] = summary_messages
+        global_stdout.append(stdout)
```
is there any specific reason for this change?

cuichenli replied:
The same reason as above

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2379,10 +2265,11 @@ def _check_directive_scope(self):
 
-    def _check_sorted_dependencies(self):
+    def _check_sorted_dependencies(
+            self, all_filepaths, global_stdout, process_manager):
```
Can we add docstring for all the params here and elsewhere?

cuichenli replied:
Will do, thanks!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2745,11 +2634,13 @@ def perform_all_lint_checks(self):
-            self._check_js_and_ts_component_name_and_count,
-            self._check_directive_scope
+            (self._check_extra_js_files, arg1),
+            (self._check_js_and_ts_component_name_and_count, arg2),
```
Do we need to pass all the files to this method?

cuichenli replied:
In this context, all files refer to js and ts files.

DubeySandeep replied:
I see, thanks for the explanation. :)

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2745,11 +2634,13 @@ def perform_all_lint_checks(self):
         common_messages = super(
             JsTsLintChecksManager, self).perform_all_lint_checks()
+        arg1 = (self.js_filepaths, _STDOUT_LIST, self.process_manager)
+        arg2 = (self.all_filepaths, _STDOUT_LIST, self.process_manager)
```
I'm still not sure why we have to pass this as an argument as it's a part of the same object? Am I missing something?

cuichenli replied:
The same reason as above

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1685,7 +1562,7 @@ def _check_codeowner_file(verbose_mode_enabled):
         python_utils.PRINT('----------------------------------------')
 
-    with _redirect_stdout(_TARGET_STDOUT):
+    with _redirect_stdout(sys.stdout):
```
Why we made this change in this place not in other places in the same file?

cuichenli replied:
That was a mistake, will fix this.

DubeySandeep reviewed (commented):
@cuichenli, I was going through this PR while fixing the linter issue and I've got a few questions for the changes in the `pre_commit_linter` file, can you please help? (I haven't build oppia in Windows before so sorry if my questions are silly.)

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -595,17 +596,16 @@
 from pylint import lint  # isort:skip
-from . import build  # isort:skip
-from . import docstrings_checker  # isort:skip
 import html.parser  # isort:skip
```
Ditto can we remove these pragmas now?

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -63,8 +63,9 @@
 import time
 
 # Install third party dependencies before proceeding.
```
is this comment required anymore? (maybe move this to the place where you have moved `install_third_party_libs.main(args=[]`.)

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -868,7 +868,7 @@ def _walk_with_gitignore(root, exclude_dirs):
         # structure.
         if (not _is_path_ignored(dir_path + '/')) and (
-                dir_path not in exclude_dirs):
+                dir_path.replace('\\', '/') not in exclude_dirs):
```
Please add a comment above (like line 868) explaining this is done for windows compatibility.

Also, have you tested this? I was expecting a similar change on line 870 as well.

seanlip replied:
Perhaps use .split() and os.path.join() instead? Ditto elsewhere you work with pathnames.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1762,13 +1639,17 @@ def _check_codeowner_file(verbose_mode_enabled):
+        file_lists = []
+        for file_pattern in file_patterns:
+            file_lists.extend(glob.glob(file_pattern))
+        target_files = {file.replace('\\', '/') for file in file_lists}
```
Why is it named target files? Is it the set of files having code owner(s)?

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1762,13 +1639,17 @@ def _check_codeowner_file(verbose_mode_enabled):
+
         for file_paths in _walk_with_gitignore('.', dir_patterns):
             for file_path in file_paths:
+                file_path = file_path.replace('\\', '/')
```
Add a comment on top, explaining this is for windows compatibility.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1762,13 +1639,17 @@ def _check_codeowner_file(verbose_mode_enabled):
-                        match = True
-                        break
+                if file_path in target_files:
+                    match = True
```
Why we are removing the `beak` statement?

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -1914,10 +1796,11 @@ def _check_mandatory_patterns(self):
         summary_messages.append(summary_message)
-        self.process_manager['mandatory'] = summary_messages
-        _STDOUT_LIST.append(stdout)
+        process_manager['mandatory'] = summary_messages
```
On line 1771 I see we are using `self` reference, if that's possible then why can't we use self.process_manager use here instead of passing it as a param? (Ditto for all_filepaths and global_stdout)

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2725,9 +2613,10 @@ def _check_dependencies(self):
         _match_line_breaks_in_controller_dependencies
         in parallel.
         """
+        arg = (self.all_filepaths, _STDOUT_LIST, self.process_manager)
```
args*

DubeySandeep replied:
Also, I think here you should only pass the ts_files.

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2745,11 +2634,13 @@ def perform_all_lint_checks(self):
         common_messages = super(
             JsTsLintChecksManager, self).perform_all_lint_checks()
+        arg1 = (self.js_filepaths, _STDOUT_LIST, self.process_manager)
+        arg2 = (self.all_filepaths, _STDOUT_LIST, self.process_manager)
```
Can we add a proper name here for args*? (If these are not used in multiple places then you can use them directly instead of creating variable.)

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -3407,11 +3308,21 @@ def main(args=None):
+    FILE_CACHE = NAME_SPACE.files
+    _FILES = multiprocessing.Manager().dict()
+
+
```
Remove extra line*

DubeySandeep reviewed (commented):
@cuichenli, Thanks for the reply, I realized you are going to create a PR for the fix so I reviewed the `pre_commit_linter` file again and left a few comments it would be great if you can cover these changes (if required) in that PR.  Thanks! :)

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1809,12 +1812,15 @@ def swap(self, obj, attr, newvalue):
         raise a RuntimeError.
         """
-        original = getattr(obj, attr)
+        original = getattr(obj, attr, NO_ATTRIBUTE)
```
I'm not really keen on this change. Conceptually, "swap" and "add/delete" are different concepts. If we really need add/delete, we should use a different contextmanager (but it would be better if we didn't need add/delete).

Why is this needed? In general, I think we should try to keep the changes in this PR minimal and incremental to avoid breaking other machines/workflows (and it's also good practice anyway).

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -210,6 +209,19 @@ def ensure_pip_library_is_installed(package, version, path):
 
+def get_yarn_command():
+    """Get command for yarn."""
+    # Rename the one without extension, otherwise it will be executed.
```
Slightly confused about what's going on here...

- Typically, "get" operations should not modify stuff. This seems to have a major side effect.
- Why will `yarn` be executed if it's not renamed?

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -1762,13 +1639,17 @@ def _check_codeowner_file(verbose_mode_enabled):
 
         # Checks that every file (except those under the dir represented by
         # the dir_patterns) is covered under CODEOWNERS.
+        file_lists = []
```
Here and elsewhere -- I don't understand why we're changing the logic, if the aim is just to get this working on Windows. Could you explain?

If we want to do both, I recommend having separate PRs for (a) logic changes, (b) Windows compatibility. That will help us isolate problems in the dev workflow PRs more effectively if any occur.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -1831,8 +1712,8 @@ def all_filepaths(self):
     def _run_multiple_checks(self, *checks):
         """Run multiple checks in parallel."""
```
I think this now needs a docstring to explain what `*checks` is (with typeinfo etc.)

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -1882,7 +1763,8 @@ def _check_for_mandatory_pattern_in_file(
 
-    def _check_mandatory_patterns(self):
+    def _check_mandatory_patterns(
+            self, all_filepaths, global_stdout, process_manager):
```
Add Args docstring.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -1986,12 +1869,15 @@ def _check_bad_patterns(self):
     def _check_patterns(self):
         """Run checks relate to bad patterns."""
-        methods = [self._check_bad_patterns, self._check_mandatory_patterns]
+        args = ((_FILES['.js'] + _FILES['.ts']), [], self.process_manager)
```
If these args are needed in the method I think they should be passed in explicitly.

seanlip commented on scripts/pre_push_hook.py:
```diff
@@ -114,7 +116,10 @@ def get_remote_name():
-    remotes = python_utils.UNICODE(out)[:-1].split('\n')
+
+    # Get rid of Windows new line symbol.
+    out = python_utils.UNICODE(out)[:-1].replace('\r', '')
```
Should only do this if platform is Windows, right?

(I know it doesn't really matter, but it seems conceptually more clean.)

Ditto elsewhere. Consider having a remove_windows_eol_characters() function somewhere if you need this to happen often.

seanlip reviewed (commented):
Hi @cuichenli -- I took a quick look at this because of the recent dev workflow breakage, and left a few comments.

Given the recent issues, I'd like to also suggest that we revert this PR and try again with all the new review comments addressed, and a more careful review to ensure that it doesn't break the existing processes. Would that be OK?

cuichenli commented:
@seanlip Feel free to revert it. I will file a series of PRs that change those files separately. But I am a little busy recently and may not be able to do those things in the short term.

seanlip commented:
OK -- thanks @cuichenli!
